hknweb/tutoring/views.py

```diff
@@ -21,5 +30,181 @@ def index(request):
+    return prefs
+
+# Generates file that will be fed into algorithm
+@permission_required('tutoring.add_slot', login_url='/accounts/login/')
+def prepare_algorithm_input(request):
+    input = {}
+    courses = []
+    for course in TutorCourse.objects.all():
+        courses.append(str(course.course))
+    input["courseName"] = courses
+    tutors = []
+    for tutor in Tutor.objects.all():
+        tutor_dict = {}
+        tutor_dict["tid"] = tutor.id
+        tutor_dict["name"] = tutor.name
+        slot_time_prefs = []
+        slot_office_prefs = []
+        slot_opposite_office_prefs = []
+        for pref in tutor.get_slot_preferences():
+            slot_time_prefs.append(pref.time_preference)
+            slot_office_prefs.append(pref.office_preference)
+            slot_opposite_office_prefs.append(-pref.office_preference)
+        tutor_dict["timeSlots"] = slot_time_prefs + slot_time_prefs
+        tutor_dict["officePrefs"] = slot_office_prefs + slot_opposite_office_prefs
+        course_prefs = []
+        for pref in tutor.get_course_preferences():
+            course_prefs.append(pref.preference)
+        tutor_dict["courses"] = course_prefs
+        tutor_dict["adjacentPref"] = tutor.adjacent_pref
+        tutor_dict["numAssignments"] = tutor.num_assignments
+        tutors.append(tutor_dict)
+    input["tutors"] = tutors
+    slots = []
+    cory_course_prefs = get_office_course_preferences(0)
+    soda_office_prefs = get_office_course_preferences(1)
+    for slot in Slot.objects.all().order_by('slot_id'):
+        slot_dict = {}
+        slot_dict["sid"] = slot.slot_id
+        slot_dict["name"] = "Slot {}".format(slot.slot_id)
+        slot_dict["adjacentSlotIDs"] = get_adjacent_slot_ids(slot.slot_id)
+        if slot.room == 0:
+            slot_dict["courses"] = cory_course_prefs
+        else:
+            slot_dict["courses"] = soda_office_prefs
+        slot_dict["day"] = slot.timeslot.get_day()
+        slot_dict["hour"] = slot.timeslot.hour
+        slot_dict["office"] = slot.get_office()
+        slots.append(slot_dict)
+    input["slots"] = slots
+    return JsonResponse(input)
+
+#Hardcoded because of the break between 3PM and 7PM for virtual semester, change back to below when in person
+def get_adjacent_slot_ids(slot_id):
+    row_interval = 2 * len(TimeSlot.DAY_CHOICES)
+    hours_threshold = (len(TimeSlot.HOUR_CHOICES) - 1) * row_interval
+    adjacent = []
+    if slot_id < row_interval:
+        adjacent.append(slot_id + row_interval)
+    elif slot_id >= hours_threshold:
+        adjacent.append(slot_id - row_interval)
+    elif slot_id >= 12 and slot_id < 18:
+        adjacent.append(slot_id - row_interval)
+    elif slot_id >= 18 and slot_id < 24:
+        adjacent.append(slot_id + row_interval)
+    else:
+        adjacent.append(slot_id - row_interval)
+        adjacent.append(slot_id + row_interval)
+    return adjacent
+
+    # row_interval = 2 * len(TimeSlot.DAY_CHOICES)
```

Comment:
Do we need this section of commented out code? If not, can we pls remove it; otherwise, can you make an issue for this and add the link in this block of comments?

Reply:
Because we have a gap in the hours, there is a version of a continuous range of time and one for (specifically) a gap between 3 to 7

[Future Reference: The tutoring desired slots for Spring 2021 is: MWF 1-3 and 7-10]

Reply:
The commented version is there as a code reference to a continuous-time block

Over time, we hopefully will make a more abstracted design later

Reply:
Can we add this information in the comment itself? (and maybe make an issue for it?)

Reply:
Added a way to automatically show it up again when two rooms are added